Rewrite LIP 0048 Rationale section (#392)
LiskHQ/lips: proposals/lip-0048.md

## Rationale

> The fee structure presented here serves to allow defining sufficient cost to manage state growth and prevent transaction pool spamming. In principle, any transaction resulting in a state change should incur a fee. This reasoning is implemented in the `payFee` method, which can be invoked by other modules.

Reviewer: I don't understand how the fee prevents transaction pool spamming, can you elaborate a bit?

Author: I'm not sure either, I believe I picked this up in some discussions related to fees. Please suggest how to rephrase it.

Author: I think I understand now what your question implied, will try to correct it.

> The fee structure presented here serves to allow defining sufficient cost to manage state growth and prevent transaction pool spamming. In principle, any transaction resulting in a state change should incur a fee. This reasoning is implemented in the `payFee` method, which can be invoked by other modules.

Reviewer: ```suggestion
The fee structure presented here serves to allow defining sufficient cost to manage state growth and prevent transaction pool spamming. In principle, any transaction resulting in a state change could pay a fee. This reasoning is implemented in the `payFee` method, which can be invoked by other modules.
```

Author: Agreed. Maybe this?
```suggestion
The fee structure presented here serves to allow defining sufficient cost to manage state growth and prevent transaction pool spamming. In principle, any transaction resulting in a state increase should pay a fee. This reasoning is implemented in the `payFee` method, which can be invoked by other modules.
```

Reviewer: Maybe I would avoid mentioning state growth altogether. Transactions should pay a fee for several reasons, e.g. state growth, computational effort from the node, bandwidth, block space, etc. What I would clearly mention is that we do not have an automatic way of calculating the fee of a transaction for each one of these costs. Instead, we only calculate the fee for the block space automatically (similarly to bitcoin if you want) with the min fee, and leave flexibility to developers to assign a fee for the other cases "by hand". For instance, in the Token module we assign by hand an initialization fee which covers for the state growth. In principle, we could also assign a fee each time an expensive calculation is required (e.g. verifying an inclusion proof).

Author: I tried to summarize your idea, please check the latest commit.

> 

Reviewer 2: I would suggest the following structure:

Subsection: Transaction fees. Include the "fee token ID" and "Minimum Fee Per Transaction" sub-sub sections.

Subsetion: CCM fees.

Then the discussion on fee pool applies to both trs and ccm fees. 

Also, we could then expalin a bit more in detail the `payFee` functionality, perhaps with one example of a transaction when it is called (e.g., validator registration) and one example of a ccm processing when it is called (e.g., need to initialize a user account).

Author: Noted.

### Fee Token ID

Each chain can configure the token used to pay fees, as defined by `TOKEN_ID_FEE`. On the Lisk mainchain, the token used for transaction fees is the LSK token.

### Minimum Fee per Transaction

> As introduced in [LIP 0013][lip-0013], all transactions must have a fee greater or equal to a minimum fee, which can be set to zero. For a transaction `trs`, the minimum fee is determined by its size and the proportionality constant `MIN_FEE_PER_BYTE`: 

Reviewer: ```suggestion
As introduced in [LIP 0013][lip-0013], all transactions must have a fee greater or equal to a minimum fee. For a transaction `trs`, the minimum fee is determined by its size and the proportionality constant `MIN_FEE_PER_BYTE`, which can be set to zero: 
```

Author: Agreed, it will be more readable that way.

```python
trsSize = length(encodeTransaction(trs))
minFee = MIN_FEE_PER_BYTE * trsSize
```

> Note that chains can freely configure their `MIN_FEE_PER_BYTE`, with a possibility to define a period of `MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE` blocks after the genesis block when it is set to zero.

Reviewer: We should add a rationale for why we allow for this. Specifically, I think it has to do with using LSK as the fee token in sidechains.

Author: Noted.

> Note that chains can freely configure their `MIN_FEE_PER_BYTE`, with a possibility to define a period of `MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE` blocks after the genesis block when it is set to zero.

Reviewer 2: ```suggestion
Note that chains can freely configure their `MIN_FEE_PER_BYTE`, which can be even set to zero. The value `MIN_FEE_PER_BYTE` can not be modified, unless a hard fork occurs. For this reason, we allow sidechains the possibility to define a period of `MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE` blocks after the genesis block for which the minimum fee does not apply. This could be useful, for example, for sidechains that want to use LSK as the fee token: initially, no LSK tokens exist in the sidechain, so the chain would not be able to have any transactions, unless `MIN_FEE_PER_BYTE = 0`. By allowing an initial exception period, users have some time window to transfer LSK tokens to this chain; after this time, `MIN_FEE_PER_BYTE` could apply with a reasonable non-zero value. 
```

Reviewer 2: This is just a suggested text and perhaps can be improved; I hope you get the main point of the suggestion.

Author: Very good, thanks for the suggestion!

### Fee Pool

> Chains can choose to either burn the minimum fee - to discourage validators from sending zero-cost transactions in the blocks they generate - or, on the other hand, to transfer it to a dedicated pool address, `ADDRESS_FEE_POOL`. 

Author: Noted.

> The latter choice could be beneficial for sidechains, as fees collected in this way can be used as an incentive for validators and other protocol participants for various purposes like funding a community pool for on-chain governance spending, sharing the funds among validators proportionally per round, incentivizing ecosystem developers, supporting chain maintenance and improvement, rewarding user staking/voting activity, etc.

Reviewer 2: perhaps we could give a more concrete example, based on DEX?  Otherwise, this paragraph looks good to me.

Author: I don't have an idea how to give a more concrete example based on DEX, please instruct me. All I know is that DEX uses the funds collected in a pool to incentivize the validators, which is already generally mentioned in the paragraph above.
Or did you have something more specific in mind?

> In the Lisk mainchain, the minimum fee is always burned.

Reviewer 2: Perhaps it would be good to give an explanation of "why" we do so (if we have a good one..). I guess one of them was to decrease the inflation, although they way we do it does not seem to have justification.

Author: Yes, I don't have a good idea. Maybe I can write something on the line of:
```suggestion
In the Lisk mainchain, the minimum fee is always burned, reducing the total supply and thus introducing a deflationary pressure to counteract the inflation.
```

### Cross-chain Message Fees

Cross-chain messages or CCMs facilitate interoperability by enabling cross-chain transfer of data. They require an associated fee, distinct from the usual transaction fee, paid by the relayer. Similarly to transaction fees, the minimum CCM fee can be either burned or transferred to a fee pool, depending on the chain configuration.

## Specification

> 

Reviewer 2: Btw I also have a comment about the verify transaction function (lines 240-255):

In line 245 we set `MIN_FEE_PER_BYTE` to 0. While I understand the intention, strictly speaking this is wrong, since we over-write the initial value.

Instead I think that what we really want to do is:

```
    if h < MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE:
	    # Set minFee to 0 for the first MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE blocks.
        minFee = 0
    else:
        trsSize = length(encodeTransaction(trs))
        minFee = MIN_FEE_PER_BYTE * trsSize
```

Author: Yes, I believe you're right. Do you think the implementation will change (I believe not)?

## Rationale

> Fees for blockchain transactions are necessary for a variety of reasons, including state growth, node computation, and block space utilization.

Reviewer: ```suggestion
Fees for blockchain transactions are necessary for a variety of reasons, including paying for state growth, node computation, and block space utilization.
```

Blockchain transaction fees are instrumental for various reasons, including: 
- Paying for state growth, node computation, and blockspace utilization. 
- Preventing spamming a blockchain with unlimited zero-cost transactions. 

> - Incentivizing validators who operate the network and maintain its security.

Reviewer 2: ```suggestion
- Incentivizing validators who operate the network and maintain its security, by providing them part of the collected fees.
```

Author: Thanks. I would only change it to:
```suggestion
- Incentivizing validators who operate the network and maintain its security, by providing them with a part of collected fees.
```

> The current implementation lacks an automated method for calculating fees in all of these cases. Instead, we automate the calculation of block space fees using a *minimum fee* approach, similar to Bitcoin's mechanism, and allow developers discretion to use the [`payFee`](#payfee) method for fee assignment in other situations, e.g., when transaction implies state growth or more significant computational effort required by a node.

Reviewer: Just to clarify: This is similar to Bitcoin in the sense that the only relevant metric for the calculation of the min fee is the transaction size, but bitcoin does not have a min fee (that I know of). For simplicity I would drop the 'similar to Bitcoin's mechanism' sentence.

Reviewer 2: Also, I would not start the paragraph by saying that we lack some feature. We could just outline the two main topics (minimum fee based on blockspace) and extra command fees and say that they are explained below.

Reviewer 2: ```suggestion
The fee system proposed here, automates the calculation of block space fees using a *minimum fee* approach: the transaction fee should exceed a threshold which depends on the size of the transaction, as we explain below. Fees related to other reasons (e.g., state growth, computational effort) are determined by the corresponding parts of the protocol (e.g., extra fee to register a validator in the PoS module); the fee module takes care of the handling of such fees, via the [`payFee`](#payfee) method.
```

Author: Thanks for the suggestion, I changed the parahraph as suggested.

> Each transaction in the Lisk ecosystem has a *transaction fee* associated with it, as defined in [LIP 0068][lip-0068]. Transaction fees prevent spamming a blockchain with unlimited transactions at zero cost. In addition, a part of the transaction fee is provided as a reward to the validators who operate the network and maintain its security, as we explain in detail below. 

Reviewer: ```suggestion
Each transaction in the Lisk ecosystem has a *transaction fee* associated with it, the `fee` property in the transaction schema in [LIP 0068][lip-0068]. Transaction fees prevent spamming a blockchain with unlimited transactions at zero cost. Part of the transaction fee is burned or transferred to a common pool. The remaining part is given as a reward to the validator who generated the block that included the transaction. 
```

Reviewer: Also, I would move this paragraph to the very top and remove this section.

Reviewer 2: Agree, this should be merged with the very first paragraph of the rationale.

Reviewer 2: Not sure we should add the burn/pool topic here. The main point of this paragraph is to explain *why* we need fees, and one of the reasons is that rewards validators. I would then keep it as is and just merge it with the very first sentence. Meaning, we just motivate the use of transaction fees.

Author: I went for the Greg's suggestion, maybe we don't need to mention the burning here, since this is moved to the Rationale introduction.

Consequently, each transaction in the Lisk ecosystem has a *transaction fee* associated with it, as defined by the `fee` property of the transaction schema in [LIP 0068][lip-0068].  

> The fee system proposed here automates the calculation of blockspace fees using a *minimum fee* approach: The transaction fee should exceed a threshold that depends on the size of the transaction. Fees associated to other factors, such as state growth or computational effort, are determined by the corresponding parts of the protocol, e.g., fee for registering a validator in the PoS module. The fee module manages these *additional* fees through the [`payFee`](#payfee) method.

Reviewer 2: ```suggestion
The fee system proposed here automates the calculation of blockspace fees using a *minimum fee* approach: The transaction fee should be greater than or equal to a threshold that depends on the size of the transaction. Fees associated to other factors, such as state growth or computational effort, are determined by the corresponding parts of the protocol, e.g., fee for registering a validator in the PoS module. The fee module manages these *additional* fees through the [`payFee`](#payfee) method.
```

> #### Fee Token ID

Reviewer: I would move this section down, after the minimum Fee section

Reviewer 2: Indeed, this is not important; we could put it in the very end of rationale for the time being, and after getting the final structure, put it in the appropriate place.

Each chain can configure the token used to pay fees, as defined by `TOKEN_ID_FEE`. On the Lisk mainchain, the token used for transaction fees is the LSK token.

#### Minimum Fee per Transaction

> As introduced in [LIP 0013][lip-0013], all transactions must have a transaction fee that is greater than or equal to a *minimum fee* that is always deducted from the part of the fee that goes to the block generator. The minimum fee, unless it is zero, is intended to discourage validators from sending zero-cost transactions in the blocks they generate.

Reviewer 2: I think also here we are confusing things. The minFee is intended to charge for blockspace. What discourages validators from sending transactions is that they don't get the minFee back in case they send a transaction in the block they generate.

Reviewer 2: ```suggestion
As introduced in [LIP 0013][lip-0013], all transactions must have a transaction fee that is greater than or equal to a *minimum fee*. This fee corresponds to a cost for using the blockspace and therefore depends on the size of the transaction.
```

Reviewer 2: Then the discussion about burning the minFee could be moved to after defining this fee.

Author: Yes, I agree that the minFee discussion is a bit unprecise; I've rewritten it according to your suggestion.

trsSize = length(encodeTransaction(trs))
minFee = MIN_FEE_PER_BYTE * trsSize
```

> The minimum fee is always subtracted from the part of the fee that goes to the block generator, as we explain below. Therefore, the minimum fee discourages validators from sending zero-cost transactions in the blocks they generate.

Reviewer 2: ```suggestion
The minimum fee is always subtracted from the part of the fee that goes to the block generator, as we explain below. This way, block generators do not have the option to include zero-cost transactions in the blocks they generate, unless minimum fee is set to zero.
```

minFee = MIN_FEE_PER_BYTE * trsSize

> ```

Reviewer 2: We can add here the discussion on removing minFee from rewards.

> ```

Reviewer 2: ```suggestion
```
The minimum fee is always deducted from the part of the fee that goes to the block generator, as we explain below. Therefore, the minimum fee, unless it is zero, discourages validators from sending zero-cost transactions in the blocks they generate.
```

Note that chains can freely configure their `MIN_FEE_PER_BYTE`, which can even be set to zero. The `MIN_FEE_PER_BYTE` value cannot be changed, unless a hard fork occurs. For this reason, we allow sidechains to define a period of `MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE` blocks after the genesis block, for which the minimum fee does not apply. 

This could, for example, be useful for sidechains that want to use LSK as the fee token. Initially, there is no LSK on the sidechain, so the sidechain cannot process transactions unless `MIN_FEE_PER_BYTE` is set to zero, which is generally not desirable since it essentially disables the feature of minimum fee for the chain. By granting an initial exemption period of `MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE` blocks, users have a window of time to transfer LSK tokens to the sidechain, after which the initially set value of `MIN_FEE_PER_BYTE` would start being applied. 

> ### Method `payFee`

Reviewer 2: I would still try to give a different title here, such as "additional fees", or "state growth fees". The payFee is just the function that does this.

Author: I renamed the paragraph "Additional Fees", I think it's a general term which doesn't say much, but covers everything :)

In principle, any transaction that results in a state increase or significant computational effort could require a fee. Since the current implementation does not provide an automatic way to charge fees in these cases, we introduce the [`payFee`](#payfee) method. 

This method can be invoked by other modules to charge an additional fee when a state increase is triggered during the execution of a transaction, e.g., when a new validator is registered in the PoS module, a new sidechain is registered, or a [user account is initialized][lip-0051#initializeuseraccountinternal] in the Token module: 
```python
Fee.payFee(USER_ACCOUNT_INITIALIZATION_FEE)
```

> In addition, developers may decide to call the `payFee` method for transactions that require extensive computation, such as when verifying an inclusion proof.

Reviewer 3: I would add something like: "..., or if any other resource is extensively used during transaction execution" to be slightly more general. Maybe a transaction accesses a decentralized storage or interacts with an oracle, etc.

Author: Right. Changed as suggested.

> For example, this method is called during the execution of the [validator registration command][lip-0057#execution] in LIP 0057: 

Reviewer: Maybe I would cite the token account initialization example rather than the validator one. The point is that in reality the validator and chain registration pay a way higher fees because you are kinda paying for the 'name space' rather than the state space. I would just avoid this discussion

Fee.payFee(VALIDATOR_REGISTRATION_FEE)
```

### Fee Processing

> A part of the fee equal to the sum of the minimum fee and, if applicable, the initialization fee, is referred to as the *consumed fee*. [After the transaction is executed](#after-command-execution), consumed fee is either burned or transferred to a dedicated pool address. 

Reviewer: I think here we should be more precise. This 'consumed fee' would be minimum fee plus other fees paid using the `payFee` method, which could be more general than initialization fees.

Reviewer 2: ```suggestion
We call the part of the fee used during the execution of a transaction (i.e., the sum of minFee plus the fees handled by payFee) as the "consumed" part of the fee.
```

Reviewer 2: Namely, I would not discuss the burn vs pool here since it destructs the flow. 

1) Define consumed, available
2) Say that available goes to validator
3) Discuss the burn vs pool.

Author: Agreed and changed as suggested.

> We refer to the total of the minimum fee and the fees processed by the `payFee` method as the "consumed" part of the fee. The remainder of the fee, called the "available" fee, is given to the validator as a reward. In this way, validators are assumed to prioritize transactions with a higher available fee in case of network congestion. [After the transaction is executed](#after-command-execution), consumed fee is either burned or transferred to a dedicated pool address. 

Reviewer 2: ```suggestion
We refer to the sum of the minimum fee and the additional fees processed by the `payFee` method as the "consumed" part of the fee. The remainder of the fee, called the "available" fee, is given to the validator as a reward. In this way, validators are assumed to prioritize transactions with a higher available fee in case of network congestion. [After the transaction is executed](#after-command-execution), consumed fee is either burned or transferred to a dedicated pool address. 
```

> The remainder of the fee, referred to as the *available fee*, is given to the block generator as a reward. In this way, validators are assumed to prioritize transactions with a higher available fee  in the case of network congestion.

Reviewer: ```suggestion
The remainder of the fee, referred to as the *available fee*, is given to the block generator as a reward. In this way, validators are assumed to prioritize transactions with a higher available fee in the case of network congestion.
```

Reviewer: Also, we already said this I think. Maybe we could put this sentence together with the paragraph in the current "### Transaction Fees" section and merge them together. Than we can remove this section and directly discuss the fee pool.

Reviewer 2: I think it is good to keep it, although it was also said before, in order to have a concrete paragraph describing this precisely.

Author: I opted for leaving it here.

> In the Lisk mainchain, the consumed fees are always burned. This reduces the total supply, creating a deflationary pressure that counteracts inflation.

Reviewer 3: I think "counteracts" is a strong word, it sounds like the burnt fees per block are comparable with minted rewards. I would say something like: "works against inflation".

Author: Changed as suggested.

> ### Cross-chain Message Fees

Reviewer 2: I think now this is the only paragraph that perhaps needs to be extended content-wise. We can discuss on this.

> Cross-chain messages, or CCMs, facilitate interoperability by enabling cross-chain transfer of data. They require an associated fee, distinct from the usual transaction fee. Similar to transaction fees, the minimum fee can either be burned or transferred to a fee pool depending on the configuration of the chain, and the available fee is rewarded to a CCM relayer.

Reviewer 3: There is no minimum fee for CCMs, only what was consumed by `payFee`, so I suggest to change "the minimum fee" to "the consumed fee".

Also I would say that the fee goes to the CCM relayer in the second sentence. It would also be good to explain that this fee is supposed to cover CCM execution and incentivize relayers.

Author: Changed as suggested.

> #### Fee Token ID

Reviewer 2: Now that we found the right structure, I think this could be added in the initial presentation of trs fees, for example after the link to LIP 68 (meaning adding this sentence there instead of having a separate sub-sub-section for it)

Author: Done.

```
The minimum fee is always subtracted from the part of the fee that goes to the block generator, as we explain below. This way, block generators do not have the option to include zero-cost transactions in the blocks they generate, unless minimum fee is set to zero.

Note that chains can freely configure `MIN_FEE_PER_BYTE`, which can even be set to zero. The `MIN_FEE_PER_BYTE` value cannot be changed, unless a hard fork occurs. For this reason, we allow sidechains to define a period of `MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE` blocks after the genesis block, for which the minimum fee does not apply. 

This could, for example, be beneficial for sidechains that want to use LSK as the fee token. Initially, there is no LSK on the sidechain, so the sidechain cannot process transactions unless `MIN_FEE_PER_BYTE` is set to zero, which is generally not desirable since it essentially disables the feature of minimum fee for the chain. By granting an initial exemption period of `MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE` blocks, users have a window of time to transfer LSK tokens to the sidechain, after which the initially set value of `MIN_FEE_PER_BYTE` would start being applied. 

> ### Additional Fees

Author: Agreed.

In principle, any transaction that results in a state increase or significant computational effort could require a fee. Since the current implementation does not provide an automatic way to charge fees in these cases, we introduce the [`payFee`](#payfee) method. 

> This method can be invoked by other modules to charge an additional fee when a state increase is triggered during the execution of a transaction, e.g., when a new validator is registered in the PoS module, a new sidechain is registered, or a [user account is initialized][lip-0051#initializeuseraccountinternal] in the Token module: 

Reviewer 2: minor: perhaps add links to validator registration or sidechain registration executions so that a reader can navigate easily and get the point.

Author: Done.

```python
Fee.payFee(USER_ACCOUNT_INITIALIZATION_FEE)
```
In addition, developers may decide to call the `payFee` method for transactions that require extensive computation, e.g., when verifying an inclusion proof, or if any other resource is extensively used during transaction execution.

> ### Fee Processing

Reviewer 2: I would still suggest to have it as a sub-sub section.

> We refer to the sum of the minimum fee and the additional fees processed by the `payFee` method as the "consumed" part of the fee. The remainder of the fee, called the "available" part of the fee, is transferred to the validator as a reward. In this way, validators are assumed to prioritize transactions with a higher available fee in case of network congestion. [After the transaction is executed](#after-command-execution), consumed fee is either burned or transferred to a dedicated pool address. 

Reviewer 2: ```suggestion
We refer to the sum of the minimum fee and the additional fees processed by the `payFee` method as the "consumed" part of the fee. The remainder of the fee, called the "available" part of the fee, is transferred to the validator as a reward. In this way, validators are assumed to prioritize transactions with a higher available fee in case of network congestion. [After the transaction is executed](#after-command-execution), consumed fees are either burned or transferred to a dedicated pool address. 
```

> #### Fee Pool

Author: Yes, seems like a good idea to remove the title, the flow will be better.

In the Lisk mainchain, the consumed fees are always burned. This reduces the total supply, creating a deflationary pressure that works against inflation.

### Cross-chain Message Fees

Cross-chain messages (CCMs) facilitate interoperability by enabling cross-chain transfer of data. Each CCM contains a [*message fee*][lip-0049#ccmschema], intended to cover the cost of transaction processing on the receiving chain (see [LIP 0049][lip-0049#fee]). 

> Since managing message fees requires checking and updating the escrow balances for the message fee token, this task is assigned to the Token module. In particular, paying for message fees is called for in the Interoperability module, whose [`sendInternal`][lip-0045#sendinternal] method uses the [`payMessageFee`][lip-0051#paymessagefee] of the Token module. After the Token module has ensured that the message fee has been credited to the relayer's account, the Fee module handles the transaction fee processing in the receiving chain.

Reviewer 2: Looks very good. One small point: I would explicitly explain that the ccm fee management includes two chains. 

e.g.., "paying for message fees *in the sending chain*", or " the ccm sender has to pay the fee in the sending chain". Then in the receiving chain, the token module assigns the fee to the relayer account and updates escrows if necessary.

Reviewer 2: Also I am not sure that so much detail of mentioning sendInternal and payMessage fee is needed, but not strong opinion.

Author: Thanks for the suggestion, I've rewritten that part.

Since managing message fees requires checking and updating the escrow balances for the message fee token, this task is assigned to the Token module. In particular, paying for message fees is called for in the Interoperability module, whose [`sendInternal`][lip-0045#sendinternal] method uses the [`payMessageFee`][lip-0051#paymessagefee] of the Token module.

> Note that, for protocol reasons (see, e.g., [`bounce`][lip-0045#bounce] and [`terminateChain`][lip-0045#terminatechain] methods), Interoperability module allows for message fees to be zero. Consequently, setting a minimum message fee should be delegated to each module that provides its own cross-chain commands. 

Author: Thanks for these suggestions, I will include them.

> Similar to transaction fees, the consumed part of the message fee can either be burned or transferred to a fee pool, depending on the chain configuration. The available part of the message fee is left to the relayer in order to cover the cost of CCM execution and to incentivize its participation in CCM relaying.

Reviewer 2: Highlight that here consumed fee is only the so-called "additional fees", since there is no minFee.

> Similar to transaction fees, the consumed part of the message fee, which might only be additional fees (as there is no mandatory minimum fee), can either be burned or transferred to a fee pool, depending on the chain configuration. The available part of the message fee is left to the relayer to cover the cost of CCM execution and to incentivize its participation in CCM relaying. However, in contrast to including transactions in a block, a relayer cannot prioritize CCMs with higher fees as the interoperability protocol requires them to be sent in order. This also implies that users have no reason to pay higher message fees than are required for successful CCM execution.

Reviewer 2: ```suggestion
The fee processing is similar to transaction fees; note that the consumed part of the message fee contains only the so-called additional fees,  as there is no mandatory minimum fee. Similarly to transaction fees, the consumed part can either be burned or transferred to a fee pool, depending on the chain configuration. The available part of the message fee is left to the relayer to cover the cost of CCU execution (note the relayer pays transaction fees for posting the CCU in the receiving chain, since a CCU is a regular transaction) and to incentivize its participation in CCM relaying. However, in contrast to including transactions in a block, a relayer cannot prioritize CCMs with higher fees as the interoperability protocol requires them to be sent in order. This also implies that users have no reason to pay higher message fees than are required for successful CCM execution.
```

Reviewer 2: Also in the very end, we could even say that (because users no need to pay more fee to include their ccm) a consequence of this is that users know that the ccms would be delivered as soon as a relayer decides to post a ccm in the target chain, avoiding this way extra delays on delivering the ccm, something like that.

The option to transfer the consumed fees to a dedicated pool address `ADDRESS_FEE_POOL` could be beneficial for sidechains, as fees collected this way can be used as an incentive for validators and other protocol participants for various purposes: Funding a community pool for on-chain governance spending, sharing funds proportionally among validators per round, incentivizing ecosystem developers, supporting chain maintenance and improvement, rewarding users' staking and voting activities, etc.

In the Lisk mainchain, the consumed fees are always burned. This reduces the total supply, creating a deflationary pressure that works against inflation.

### Cross-chain Message Fees

Cross-chain messages (CCMs) facilitate interoperability by enabling cross-chain transfer of data. Each CCM contains a [*message fee*][lip-0049#ccmschema], intended to cover the cost of transaction processing on the receiving chain (see [LIP 0049][lip-0049#fee]). 

> Since managing message fees requires checking and updating the escrow balances for the message fee token, this task is assigned to the Token module whose [`payMessageFee`][lip-0051#paymessagefee] method charges message fees in the sending chain. In the receiving chain, the Token module assigns the message fee to the relayer account and updates escrows if necessary, and then the Fee module handles the transaction fee processing as described above. Note that the CCM sender should include any additional fees necessary for the CCM execution, e.g., to pay for the user account initialization on the receiving chain.

Reviewer 2: The last sentence of this paragraph, could be moved to the end of the next paragraph IMO

Author: Yes, I noticed it doesn't fit so well here. Thanks for the suggestion.

Reviewer 2: You forgot to remove it from here, you just pasted in the next paragraph ;-)

> Since managing message fees requires checking and updating the escrow balances for the message fee token, this task is assigned to the Token module whose [`payMessageFee`][lip-0051#paymessagefee] method charges message fees in the sending chain. In the receiving chain, the Token module assigns the message fee to the relayer account and updates escrows if necessary, and then the Fee module handles the transaction fee processing as described above. Note that the CCM sender should include any additional fees necessary for the CCM execution, e.g., to pay for the user account initialization on the receiving chain.

Reviewer 2: ```suggestion
Since managing message fees requires checking and updating the escrow balances for the message fee token, this task is assigned to the Token module whose [`payMessageFee`][lip-0051#paymessagefee] method charges message fees in the sending chain. In the receiving chain, the Token module assigns the message fee to the relayer account and updates escrows if necessary, and then the Fee module handles the message fee processing in a similar way as it handles transaction fees.
```

Author: Oh. Thanks for noticing this

#### Minimum Fee per CCM

For protocol reasons (see, e.g., [`bounce`][lip-0045#bounce] and [`terminateChain`][lip-0045#terminatechain] methods in LIP 0045), the Interoperability module allows the message fee to be zero. Consequently, the choice of setting a minimum message fee is delegated to each module that provides its own cross-chain commands. Nevertheless, there is a minimum fee in case the user wants the failed CCM to be returned (see [`bounce`][lip-0045#bounce] method in LIP 0045). Thus, even though this fee is not mandatory, it is advisable to treat it as a suggested minimum fee. 

#### Message Fee Processing

> Processing of message fees is similar to that of transaction fees. It should be noted that the consumed part of the message fee contains only the so-called additional fees, as there is no mandatory minimum fee. The consumed part can either be burned or transferred to a fee pool, depending on the chain configuration. 

Reviewer 2: ```suggestion
Processing of message fees is similar to that of transaction fees. The only essential difference is that the consumed part of the message fee contains only the so-called additional fees, as there is no mandatory minimum fee. The consumed part is handled in the same way as for transactions (i.e., either be burned or transferred to a fee pool, depending on the chain configuration) . 
```

Reviewer 2: This just to make clear that consumed fee processing is the same for transactions and ccms, for example it is not possible to burn the transaction fee and send the ccm fee to a pool.

> The available part of the message fee is left to the relayer to cover the cost of CCU execution and to incentivize its participation in CCM relaying. Also, since a CCU is a regular transaction, note that the relayer pays transaction fees for posting the CCU in the receiving chain. 

Reviewer 2: ```suggestion
The available part of the message fee is left to the relayer to cover the cost of CCU execution (recall that a CCU is a regular transaction, therefore the relayer has to pay transaction fees for posting the CCU in the receiving chain) and to incentivize its participation in CCM relaying. 
```

Reviewer 2: Suggest this change because the last sentence seemed out of context compared to the first one.

## Rationale

> ### Initialization Fee

Reviewer 2: In my opinion, starting with this topic is a bit confusing. Moreover, not all extra-command fees are really initialization fees, for example the sidechain registration fee. 

I would suggest to start with the transaction fee subsection. 

We could even link to the transaction schema LIP and start by saying that each transaction has a fee.

Reviewer 2: Suggested structure (unpolished):

Subsection: Transaction Fees.

In the lisk ecosystem, as in almost all blockchains, each transaction has an associated transaction fee (link transaction schema LIP 68). Transaction fees prevent spamming a blockchain with unlimited transactions at no cost. Moreover, part of the transaction fees is provided as reward to validators who run and maintain the security of the network, as we explain in detail below.  [this part makes clear *why* we have transaction fees].

[Then add the sub-subsections of FeeTokenID and Minimum Fee Per Transaction, which are very well written IMO. ]

Subsection: Initialization fees ( we could even call them state growth fees or something else)

The fee structure presented here allows for defining a sufficient cost to manage state growth. In principle, any transaction resulting in a state increase should imply a fee.  This reasoning is implemented in the [`payFee`](#payfee) method, which can be invoked by other modules to charge an additional fee if during the execution of a transaction, a state growth is trigerred (e.g., new validator register in PoS module, new sidechain gets registered, or initialize a new user account).

[then discuss what happens after trs execution]

We call the part of the fee used during the execution of a transaction (i.e., the sum of minFee plus the initialization fees) as the "consumed" part of the fee;  the rest of the fee is called available fee and it is given as reward to the block generator, as an additional reward for securing the network. This way, it is expected that whenever the network is congested, validators prioritize transactions with higher available fee. 

[Add here the Fee pool paragraph]

[Add the ccm paragraph]

Author: Yes, I also noticed that starting with "initialization" fees (which are, I agree, not only that, so the naming is not perfect) is a bit awkward, but I had no idea where to introduce them so I thought the beginning of the Rationale section is OK. 
Thanks for the suggestions, it's better this way since transaction fees are discussed first, which is a more natural thing to do.

The fee structure presented here allows for defining a sufficient cost to manage state growth. In principle, any transaction resulting in a state increase should imply a fee. This reasoning is implemented in the [`payFee`](#payfee) method, which can be invoked by other modules to charge *initialization fee* for data store initializations or some other special purposes, e.g., chain or validator registration. 

For example, this method is called during the execution of the [validator registration command][lip-0057#execution] in LIP 0057: 
```python
Fee.payFee(VALIDATOR_REGISTRATION_FEE)
```

### Transaction Fee

*Transaction fees* are introduced to motivate validators. Any remaining amount from the transaction fee, not consumed by the transaction processing, is given as a reward to the validator that includes the transaction in a block. 

#### Fee Token ID

Each chain can configure the token used to pay fees, as defined by `TOKEN_ID_FEE`. On the Lisk mainchain, the token used for transaction fees is the LSK token.

#### Minimum Fee per Transaction

> As introduced in [LIP 0013][lip-0013], all transactions must have a transaction fee greater or equal to a *minimum fee*, which is introduced in order to discourage validators from sending zero-cost transactions in the blocks they generate. 

Reviewer 2: At this stage of the text, perhaps it is not so clear why this discourages validators. Perhaps it would be good to add a sentence explaining this. 

Example: 

As introduced in [LIP 0013][lip-0013], all transactions must have a transaction fee greater or equal to a *minimum fee*. This is always deducted from the part of the fee which is given to the block generator. This way, the minimum fee (if non zero) is designed to discourage validators from sending zero-cost transactions in the blocks they generate.

Author: Agreed, this makes it more clear. The nuances...

For a transaction `trs`, the minimum fee is determined by its size and the proportionality constant `MIN_FEE_PER_BYTE`: 
```python
trsSize = length(encodeTransaction(trs))
minFee = MIN_FEE_PER_BYTE * trsSize
```
Note that chains can freely configure their `MIN_FEE_PER_BYTE`, which can be even set to zero. The value `MIN_FEE_PER_BYTE` can not be modified, unless a hard fork occurs. For this reason, we allow sidechains the possibility to define a period of `MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE` blocks after the genesis block for which the minimum fee does not apply. 

> This could be useful, for example, for sidechains that want to use LSK as the fee token: Initially, no LSK is present on the sidechain, so the sidechain would not be able to process any transactions unless `MIN_FEE_PER_BYTE` is set to zero. By allowing an initial exception period, users have a time window to transfer LSK tokens to this sidechain, after which the initially set value of `MIN_FEE_PER_BYTE` would start being applied. 

Author: Noted.

    b = block including trs
    h = b.header.height
    if h < MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE:

> 	    # Set MIN_FEE_PER_BYTE = 0 for the first MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE blocks.

Author: ```suggestion
	            # Set MIN_FEE_PER_BYTE = 0 for the first MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE blocks.
```

>     # Set minFee = 0 for the first MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE blocks.

Author: ```suggestion
	                    # Set MIN_FEE_PER_BYTE = 0 for the first MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE blocks.
```

    b = block including trs
    h = b.header.height
    if h < MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE:

>                                     # Set MIN_FEE_PER_BYTE = 0 for the first MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE blocks.
> 	                            minFee = 0
>         minFee = 0

Author: ```suggestion
        # Set MIN_FEE_PER_BYTE = 0 for the first MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE blocks.
        minFee = 0
```

>     b = block including trs

Reviewer 2: Comment about line 249: Perhaps add a sentence or two explaining this? In particular, that this module runs this step before *every* transaction submitted in the chain (perhaps link to LIP55 defining the steps..)

Reviewer 2: Minor point: I meant to add it as text between subsection title and pseudocode. THis way you could put also a link to LIP 55).

    h = b.header.height
    if h < MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE:

> 	    # Set minFee = 0 for the first MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE blocks.

Reviewer 2: indentation is off.

Author: I don't know what is happening with my VSC, I'll commit from this suggestion:
```suggestion
    # Set minFee = 0 for the first MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE blocks.
```

Author: Second try:
```suggestion
        # Set minFee = 0 for the first MAX_BLOCK_HEIGHT_ZERO_FEE_PER_BYTE blocks.
```
